Campaign screen final

ottml opened the pull request:
Fixes https://github.com/Return-To-The-Roots/s25client/issues/568
Relates too #1591

ottml commented:
![romancampaign](https://github.com/Return-To-The-Roots/s25client/assets/33401986/933070dc-63c3-45d2-8c13-49857d6748a0)
![kontinetcampaign](https://github.com/Return-To-The-Roots/s25client/assets/33401986/2b7fc8e1-d36c-4863-a881-d198d5920d39)

Spikeone commented:
Looking pretty good - although it's sad we don't have the original worldmap screen, I don't think we need that (for now).

Spikeone commented:
Oh, and we still need some kind of "patch map" function where one is able to override parts of the original map file (to fix BBs maps) - since this is only required for the original roman campaign, it could be just giving an index and new hex value which changes the mapfile after initially loading it.

ottml commented:
@Flamefire @Flow86 can you have a look at the pull request and make a review. Would be nice. Thanks.

Flamefire reviewed (changes requested):
Awesome work!

A bit of cleanup would be great and I did my best to find potential issues and suggest solutions & improvements.

Spikeone commented on data/RTTR/CAMPAIGNS/world/AFRICA.LUA:
```diff
@@ -0,0 +1,68 @@
+
+-------------------------------- Lua Version used -----------------------------
+function getRequiredLuaVersion()
+    return 1.3
```
I think it should be 1 instead of 1.3, guess 3 was the feature level which has been removed (in all scripts, also the roman version)

Spikeone commented on data/RTTR/CAMPAIGNS/roman/MISS200.LUA:
```diff
@@ -0,0 +1,514 @@
+    return 1.3
+end
+
+
```
Guess all roman scripts require the preview disabled, haven't done that yet

Spikeone commented:
@Flamefire shouldn't we put campaigns into the s25maps repo as well?

Flamefire commented on libs/rttrConfig/src/files.h:
```diff
@@ -35,6 +35,7 @@ namespace folders {
     constexpr auto mapsPlayed = "<RTTR_USERDATA>/MAPS"; // downloaded maps
     constexpr auto mapsRttr = "<RTTR_RTTR>/MAPS/NEW";   // maps added by RTTR
     constexpr auto mapsSea = "<RTTR_RTTR>/MAPS/SEA";    // seafaring maps
+    constexpr auto campaigns = "<RTTR_RTTR>/CAMPAIGNS"; // contains one folder for each campaign
```
Should we maybe have a user folder for campaigns, e.g. `<RTTR_USERDATA>/campaigns`? The `RTTR_RTTR` folder is in the install dir and might not be user-writable.

Flow86 replied:
perhaps load both?
the first one in `RTTR_RTTR` for our campaigns (aka romans, ...) and for users campaigns one in `RTTR_USERDATA` - perhaps like we have with the map selection with categories?

ottml replied:
Added the additional search path for user defined campaigns.

Flamefire commented:
> @Flamefire shouldn't we put campaigns into the s25maps repo as well?

I'd say it makes sense to have "our" campaigns here. At least the lua files. Otherwise we'd need another submodule and as LUA files are text they don't hurt to have in this repo. Only binary files like maps should be separate. At least that was the intention for the maps repo IIRC.

Spikeone commented:
> > @Flamefire shouldn't we put campaigns into the s25maps repo as well?
> 
> I'd say it makes sense to have "our" campaigns here. At least the lua files. Otherwise we'd need another submodule and as LUA files are text they don't hurt to have in this repo. Only binary files like maps should be separate. At least that was the intention for the maps repo IIRC.

So, since I have a custom campaign which was made by a fan, we then init that one? Fine for me, was just thinking about easy PRs, even for the campaign maps

Flamefire commented:
> So, since I have a custom campaign which was made by a fan, we then init that one? Fine for me, was just thinking about easy PRs, even for the campaign maps

Makes sense I guess. No need to run the full CI suite for campaign PRs. On the other hand: We need a PR here for the submodule update then which results in the same. But with the (potential) maps it makes sense to pull it out. I guess we'll need a new repo for that. @Flow86 ?

However I wouldn't block the PR for that: Let's get this in first, moving the text files into a subproject can be done afterwards.

ottml commented on libs/s25main/desktops/dskCampaignSelection.cpp:
```diff
@@ -0,0 +1,242 @@
+    ID_NEXT
+};
+
+int startOffsetY = 20;
```
Done

ottml commented on libs/s25main/desktops/dskCampaignSelection.cpp:
```diff
@@ -0,0 +1,242 @@
+int firstColumnExtentX = 480;
+int secondColumnExtentX = 280;
+int columnExtentY = 500;
+int scecondColumnOffsetX = padding * 2 + firstColumnExtentX;
```
Done

Flamefire commented:
Please don't force-push during a review phase. It is difficult to find what you have changed, i.e. what parts of the review comments were adressed, after a force-push. You can commit with `git commit --fixup=<original commit sha>` to note that this previous commit should be fixed/changed and do that automatically after the review/before merge with `git rebase --autosquash`

ottml commented:
> Please don't force-push during a review phase. It is difficult to find what you have changed, i.e. what parts of the review comments were adressed, after a force-push. You can commit with `git commit --fixup=<original commit sha>` to note that this previous commit should be fixed/changed and do that automatically after the review/before merge with `git rebase --autosquash`

Thanks for the  hint. I will do this from now on. I do not  know that this is possible in such an esay was.

Flamefire commented on libs/s25main/QuickStartGame.cpp:
```diff
@@ -87,8 +87,10 @@ bool QuickStartGame(const boost::filesystem::path& mapOrReplayPath, const std::v
+    if((extension == ".sav"
+        && GAMECLIENT.HostGame(csi, mapOrReplayPath, boost::optional<boost::filesystem::path>(), MapType::Savegame))
+       || ((extension == ".swd" || extension == ".wld")
+           && GAMECLIENT.HostGame(csi, mapOrReplayPath, boost::optional<boost::filesystem::path>(), MapType::OldMap)))
```
Use `boost::none` for an empty optional which is much easier to read:

```suggestion
           && GAMECLIENT.HostGame(csi, mapOrReplayPath, boost::none, MapType::OldMap)))
```

Flamefire commented:
> Thanks for the hint. I will do this from now on. I do not know that this is possible in such an easy way.

No worries, that seems to be something new to git as I learned about that only recently. Previously I did that manually with stuff like `git commit -m 'f: Part of another commit title'` and then do the fixups with an interactive rebase manually (`git rebase -i`)

And with that change from GitHub showing the diff of a force-push it actually got easier to even track those.

Flow86 commented on data/RTTR/CAMPAIGNS/roman/MISS200.LUA:
```diff
@@ -0,0 +1,517 @@
+        return
+    end
+
+    if(     (x == 34) and (y == 28) ) then MissionEvent(4) -- RttR calls onOccupied once a pole is on that spot, SettlersII called the function once the spot is really owned
```
> RttR calls onOccupied once a pole is on that spot, SettlersII called the function once the spot is really owned

what do you mean by that? whats exactly the difference?

ottml replied:
I copied the lua files from @Spikeone repositiory. The difference is here when the border is on the position the mission is won and in the original settler the mission is won when the gate in witgin the teritory. I don't think this is a problem for now.

Flow86 commented on data/RTTR/CAMPAIGNS/roman/campaign.lua:
```diff
@@ -0,0 +1,29 @@
+rttr:RegisterTranslations(
```
do we have something like `getRequiredLuaVersion` for campaign luas too? that would be good to add something like that from the start, so we can add newer features later easily

ottml replied:
Introduced

Flow86 commented on libs/s25main/network/GameClient.cpp:
```diff
@@ -54,6 +54,23 @@
 #include <memory>
 
+namespace {
+void copyFileIfPathDifferent(const boost::filesystem::path& src_path, const boost::filesystem::path& dst_path)
```
we should put this function somewhere like libutil or so

ottml replied:
Can we do this in another merge request or is there a location where i can move the function without having to update a submodule. If i understand the submodules correct i have to make a PR for the submodule repo and then after this is merged back i have to change the pointer to the submodule in a PR of the s5client repo.

Flamefire replied:
Yes, IMO we can do that after this PR

Flow86 commented on libs/s25main/network/GameClient.h:
```diff
@@ -80,7 +81,8 @@ class GameClient final :
                  bool host, bool use_ipv6);
     /// Start the server and connect to it
-    bool HostGame(const CreateServerInfo& csi, const boost::filesystem::path& map_path, MapType map_type);
+    bool HostGame(const CreateServerInfo& csi, const boost::filesystem::path& map_path,
```
somehow having so many parameters in the mean time it should probably be best to use a struct for it:
```suggestion
struct MapInfo
{
boost::filesystem::path map_path;
MapType map_type;
boost::optional<boost::filesystem::path lua_path;
};
bool HostGame(const CreateServerInfo& csi, const MapInfo& map);
```

so we don't need to use `boost::none` but can use struct initialization.

ottml replied:
Use name MapDescription because MapInfo already exists for other parameters.

Flow86 commented on libs/s25main/network/GameServer.cpp:
```diff
@@ -185,7 +186,7 @@ bool GameServer::Start(const CreateServerInfo& csi, const boost::filesystem::pat
 
-    bfs::path luaFilePath = bfs::path(mapinfo.filepath).replace_extension("lua");
+    bfs::path luaFilePath = lua_path.get_value_or(bfs::path(mapinfo.filepath).replace_extension("lua"));
     if(bfs::is_regular_file(luaFilePath))
```
shouldn't it be an error if we explicitly want to load a lua file, but it does not exist? (e.g `lua_path` *not* empty but file does not exist)

Flamefire commented on libs/libGamedata/lua/CampaignDataLoader.cpp:
```diff
@@ -38,14 +40,7 @@ bool CampaignDataLoader::Load()
-void CampaignDataLoader::AddCampaign(const kaguya::LuaTable& data)
-{
-    CampaignDescription(data, campaignDesc_);
+    state["RTTRCampaignData"].setClass(kaguya::UserdataMetatable<CampaignDataLoader, LuaInterfaceBase>());
```
We don't need that Register now, do we?

ottml replied:
I needed this otherwise it does not work. But i am not a lua expert therefore i am not sure what this line is doing really. In my understanding this is needed to get the connection betweeen lua and our class. Any input is welcome :)

Flamefire commented on libs/libGamedata/gameData/CampaignDescription.cpp:
```diff
@@ -1,28 +1,42 @@
-CampaignDescription::CampaignDescription() = default;
-
-CampaignDescription::CampaignDescription(CheckedLuaTable luaData, CampaignDescription& campaignDesc)
+CampaignDescription initFromLuaTable(kaguya::LuaTable table)
```
This now looks like it can become a regular (and `explicit`) constructor, doesn't it? I mean you create and return a new instance which is basically what constructors are for. Might be even cleaner as you can use an initializer list and only do the validity-checks on the body.

ottml replied:
Changed to constructor

Flamefire commented on libs/libGamedata/lua/CampaignDataLoader.cpp:
```diff
@@ -28,8 +28,10 @@ bool CampaignDataLoader::Load()
     try
     {
-        if(!loadScript(curFile_))
+        if(!loadScript(curFile_) || lua["campaign"].isNilref() || lua["campaign"].type() != LUA_TTABLE)
```
Maybe just get the campaign as a lua ref once instead of 3 times? See e.g. https://github.com/Return-To-The-Roots/s25client/blob/29dc07bfb32d96c0e931ddffa5d4d4a39419549c/libs/libGamedata/lua/LuaInterfaceBase.cpp#L109-L110 and IIRC `type() == LUA_TTABLE` implies `!isNilref()`
You could even throw an exception (caught below) if the entry does not exist to log that `campaign` is missing

ottml replied:
Changed

Flamefire commented on libs/libGamedata/lua/CampaignDataLoader.cpp:
```diff
@@ -28,8 +28,10 @@ bool CampaignDataLoader::Load()
+        if(!loadScript(curFile_) || lua["campaign"].isNilref() || lua["campaign"].type() != LUA_TTABLE)
             return false;
+
+        campaignDesc_ = initFromLuaTable(lua["campaign"]);
```
With the above suggestion:

```suggestion
        campaignDesc_ = CampaignDescription(lua["campaign"]);
```

ottml replied:
Done

Flamefire commented on tests/s25Main/campaign/testCampaignLuaFile.cpp:
```diff
@@ -1,16 +1,27 @@
 #include <boost/test/unit_test.hpp>
 
+/// Require that the log contains "content"
+#define REQUIRE_LOG_CONTAINS(content)                                                                                  \
```
Don't we have the macro already defined somewhere?

ottml replied:
We have only a macro defined, which checks the first line of the log and i have to check for content of multiple lines.

Flamefire commented on tests/s25Main/campaign/testCampaignLuaFile.cpp:
```diff
@@ -0,0 +1,211 @@
+    CampaignDataLoader loader(desc, tmp.get());
+    rttr::test::LogAccessor logAcc;
+    BOOST_TEST(!loader.Load());
+    REQUIRE_LOG_CONTAINS(
+      "Failed to load campaign data!\nReason: Failed to load game data: Required field 'luaFolder' not found");
```
Maybe just throw a `std::runtime_error` such that this reads:
```suggestion
    REQUIRE_LOG_CONTAINS("Failed to load campaign data!\nReason: Required field 'luaFolder' not found");
```

ottml replied:
Thats not possible because CheckedLuaTable throws this error internal.

Flamefire commented on libs/s25main/network/GameClient.h:
```diff
@@ -79,8 +80,16 @@ class GameClient final :
     bool Connect(const std::string& server, const std::string& password, ServerType servertyp, unsigned short port,
                  bool host, bool use_ipv6);
+
+    struct MapDescription
```
This should be used by the GameServer too, doesn't it? Maybe move it to a new file in `gameTypes`

ottml replied:
Yes i thought about this. I can do it.

ottml replied:
done

ottml commented on libs/s25main/desktops/dskCampaignSelection.cpp:
```diff
@@ -0,0 +1,242 @@
+            if(int ec = libsiedler2::loader::LoadMAP(mapPath, map, true))
+            {
+                LOG.write(_("Failed to load map %1%: %2%\n")) % mapPath % libsiedler2::getErrorString(ec);
+                brokenMapPaths.insert(folder);
```
Remove brokenMapPaths completly.

ottml commented on libs/s25main/desktops/dskCampaignSelection.cpp:
```diff
@@ -0,0 +1,242 @@
+    // replace this by AddImageButton as soon as it can handle this sceneraio correctly
+    if(showCampaignInfo_ && campaignImageId_)
+    {
+        auto* image = LOADER.GetImageN(*campaignImageId_, 0);
```
If i remember correct the pull request #1536 is needed, to place the image correct on the ctrlImage button.

Flamefire commented on libs/s25main/desktops/dskCampaignSelection.cpp:
```diff
@@ -0,0 +1,246 @@
+
+    showCampaignInfo(false);
+
+    AddTimer(ID_TimerFillCampaignTable, 1ms);
```
Why the timer instead of a direct call?

ottml replied:
We need the timer here because otherwise the error message showing the campaigns not loaded succesful is not displayed. Because we are in the constructor of the class and the call is not processed.

Flamefire commented on libs/s25main/desktops/dskCampaignSelection.cpp:
```diff
@@ -0,0 +1,246 @@
+                }
+            }
+
+            if(success)
```
This can never be false at this point, as you always `continue` where you set this to false, or am I missing anything?

ottml replied:
The continue effects only the inner loop over the maps. So if any error during check of the maps occours success will be false after the inner loop has finished. The continue is only there to avoid checking further stuff for an already failed. But i want to log all errors for all maps and not only the first failed map.

ottml commented:
> @ottml I updated the script files sligthly, maybe you can also update them here (https://github.com/Spikeone/RttR_Campaigns/tree/master/Campaign_OriginalRoman) - they now use your mappreview function (although this could have been a campaign setting, thinking about it)

Updated to your version.

Flamefire commented on libs/s25main/desktops/dskCampaignMissionSelection.cpp:
```diff
@@ -0,0 +1,197 @@
+        if(missionIndex >= settings_->mapNames.size())
+            break;
+
+        const bfs::path& mapFilePath = RTTRCONFIG.ExpandPath(settings_->mapFolder) / settings_->mapNames[missionIndex];
```
At least use the function (and get rid of the RTTRConfig include):
```suggestion
        const bfs::path& mapFilePath = settings_->getMapFilePath(settings_->mapNames[missionIndex]);
```

That's also why I thought about making `mapNames/luaFolder/mapFolder` private and provide only access to the map count and map/lua file path by index. Iterating over it (instead of mapNames) could be done by `for(auto idx: helpers::range(settings.getNumMaps()))` but here you already have indices

ottml replied:
Potential i will implement this :)

ottml replied:
Added index access and made mapNames/luaFolder/mapFolder private. And use it at calling side.

Flamefire commented on libs/s25main/desktops/dskCampaignSelection.cpp:
```diff
@@ -0,0 +1,246 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#include "dskCampaignSelection.h"
+#include "Desktop.h"
+#include "ListDir.h"
+#include "Loader.h"
+#include "RttrConfig.h"
+#include "Window.h"
```
```suggestion
#include "ListDir.h"
#include "Loader.h"
#include "RttrConfig.h"
```

Flamefire commented on libs/s25main/desktops/dskCampaignSelection.cpp:
```diff
@@ -0,0 +1,246 @@
+constexpr int firstColumnExtentX = 480;
+constexpr int secondColumnExtentX = 280;
+constexpr int columnExtentY = 500;
+constexpr int scecondColumnOffsetX = padding * 2 + firstColumnExtentX;
```
```suggestion
constexpr int secondColumnOffsetX = padding * 2 + firstColumnExtentX;
```

Flamefire commented on libs/s25main/desktops/dskCampaignSelection.cpp:
```diff
@@ -0,0 +1,246 @@
+{
+    auto* table = GetCtrl<ctrlTable>(ID_Table);
+    const auto& selection = table->GetSelection();
+    boost::filesystem::path campaingPath = selection ? table->GetItemText(*selection, 5) : boost::filesystem::path();
+    if(campaingPath.empty())
+        throw std::runtime_error("Campaign path is empty.");
+    return campaingPath;
```
The column can never be empty, can it?
Maybe rather:
```suggestion
    if(!selection)
        throw std::runtime_error("Campaign path is empty.");
    boost::filesystem::path campaignPath = table->GetItemText(*selection, 5);
    Rttr_Assert(!campaignPath.empty());
    return campaignPath;
```

Flamefire commented on libs/s25main/desktops/dskCampaignSelection.h:
```diff
@@ -0,0 +1,33 @@
+    bool showCampaignInfo_;
+    CreateServerInfo csi_;
+    glArchivItem_Bitmap* campaignImage_;
+    /// Campaings that we already know are broken
```
```suggestion
    /// Campaigns that we already know are broken
```

Flamefire commented on tests/s25Main/campaign/testCampaignLuaFile.cpp:
```diff
@@ -0,0 +1,274 @@
+    BOOST_TEST(desc.maxHumanPlayers == 1u);
+    BOOST_TEST(desc.difficulty == "easy");
+    BOOST_TEST(desc.mapFolder == "<RTTR_GAME>/DATA/MAPS");
+    BOOST_TEST(desc.luaFolder == "<RTTR_GAME>/CAMPAINGS/ROMAN");
```
```suggestion
    BOOST_TEST(desc.luaFolder == "<RTTR_GAME>/CAMPAIGNS/ROMAN");
```

Flamefire commented on libs/s25main/network/GameClient.h:
```diff
@@ -13,12 +13,14 @@
 #include "gameTypes/TeamTypes.h"
 #include "gameTypes/VisualSettings.h"
 #include "s25util/Singleton.h"
+#include <boost/optional.hpp>
```
```suggestion
```

Flamefire commented on libs/s25main/network/GameServer.h:
```diff
@@ -10,11 +10,13 @@
 #include "liblobby/LobbyInterface.h"
 #include "s25util/LANDiscoveryService.h"
 #include "s25util/Singleton.h"
+#include <boost/optional.hpp>
```
```suggestion
```

Flamefire reviewed (changes requested):
Looking good, almost done. Found a couple minor things

The map/lua-path-by-index thing is strictly optional. The other things are mostly spelling issues/typos.

I expect this to be ready afterwards. Can you provide screenshots of the 2 new screens in this PR? @Spikeone please check them :)

Thanks again!

ottml reviewed (commented):
Added suggestions.

Flamefire commented on doc/AddingCustomCampaign.md:
```diff
@@ -0,0 +1,94 @@
+
+## Location for adding a new campaign
+
+To add a new custom campaign you have to create a new subfolder under `RTTR/CAMPAIGNS/` or under `<RTTR_USERDATA>/CAMPAIGNS/` with your campaign name. This subfolder will be filled with all the needed data for the campaign.
```
I'm afraid users won't understand this without an example:
```suggestion
To add a new custom campaign you have to create a new subfolder with your campaign name under `RTTR/CAMPAIGNS/` in the install directory or under `<RTTR_USERDATA>/CAMPAIGNS/`.
`<RTTR_USERDATA>` is the placeholder for the RttR folder in your user dir, e.g. `~/.s25rttr` on Linux or "Saved Games" on Windows.  
This subfolder will be filled with all the needed data for the campaign.
```

Flamefire reviewed (changes requested):
Just a minor thing, besides that it is good to go from my side :+1: 

Please rebase (with `--autosquash` to incorporate the fixups) and force push.   
You'll likely need to wait for https://github.com/Return-To-The-Roots/s25client/pull/1634 though for the fix of CI after something has changed in the runners.

Flamefire commented on libs/libGamedata/gameData/CampaignDescription.cpp:
```diff
@@ -37,12 +37,28 @@ CampaignDescription::CampaignDescription(const kaguya::LuaRef& table)
+{
+    if(idx < mapNames.size())
+        return (RTTRCONFIG.ExpandPath(luaFolder) / mapNames[idx]).replace_extension("lua");
+    return boost::filesystem::path();
```
Better throw an exception here and in the others

ottml replied:
Make no check anymore at use at(idx) on vector container so a exception is thrown automatically when out of range.

ottml commented:
![campaign_final1](https://github.com/Return-To-The-Roots/s25client/assets/33401986/85780265-9269-4d0f-8225-0464e709d055)
![campaign_final2](https://github.com/Return-To-The-Roots/s25client/assets/33401986/c38bed67-3f9b-460e-a33d-33b90069f5f8)
![campaign_final3](https://github.com/Return-To-The-Roots/s25client/assets/33401986/bd2f71db-cb20-4d99-9757-929a3f497782)
![campaign_final4](https://github.com/Return-To-The-Roots/s25client/assets/33401986/e8947f0e-4eb6-4116-9ae6-d749c28d52fb)
![campaign_final5](https://github.com/Return-To-The-Roots/s25client/assets/33401986/aa029f6e-a822-4189-8272-44915c8a6763)

Flamefire commented:
Awesome, I think this is ready now!

@ottml I rebased the branch squashing the fixup-commits into their originals (i.e `git commit -i --autosquash master`) and added a minor improvement on top to avoid you having to do another change after you already did so much here. I hope that is ok for you, otherwise feel free to do the squashing-rebase yourself and force-push your branch again.   
Note: If you want/need to work on the updated (rebased) branch again you'll have the update your local branch: `git fetch <your-remote> && git reset --hard <your-remote>/campaign_screen_final`. But with your agreement I'll merge this as soon as CI has passed so you can also just delete that local branch and continue of the master after the merge.    

Please tell me if this is OK for you to merge this with that. It is your work so I'm waiting for your OK or update here. Thank you!

ottml commented:
> Awesome, I think this is ready now!
> 
> @ottml I rebased the branch squashing the fixup-commits into their originals (i.e `git commit -i --autosquash master`) and added a minor improvement on top to avoid you having to do another change after you already did so much here. I hope that is ok for you, otherwise feel free to do the squashing-rebase yourself and force-push your branch again. Note: If you want/need to work on the updated (rebased) branch again you'll have the update your local branch: `git fetch <your-remote> && git reset --hard <your-remote>/campaign_screen_final`. But with your agreement I'll merge this as soon as CI has passed so you can also just delete that local branch and continue of the master after the merge.
> 
> Please tell me if this is OK for you to merge this with that. It is your work so I'm waiting for your OK or update here. Thank you!

@Flamefire It's OK for me to merge it. Thanks for the fast and productive review :)

Flamefire commented:
@Flow86 Blocked by:

- https://github.com/Return-To-The-Roots/s25client/pull/1634
